src/shogun/machine/gp/SingleFITCLaplacianInferenceMethod.cpp

```diff
@@ -0,0 +1,956 @@
+{
+	SG_DEBUG("entering\n");
+
+	CInferenceMethod::update();
+	update_init();
+	update_alpha();
+	update_chol();
+	update_approx_cov();
+	update_deriv();
+	update_parameter_hash();
+
+	SG_DEBUG("leaving\n");
+}
+
+SGVector<float64_t> CSingleFITCLaplacianInferenceMethod::get_diagonal_vector()
+{
+	if (parameter_hash_changed())
+		update();
+
+	return SGVector<float64_t>(m_sW);
+}
+
+CSingleFITCLaplacianInferenceMethod::~CSingleFITCLaplacianInferenceMethod()
+{
+}
+
+SGVector<float64_t> CSingleFITCLaplacianInferenceMethod::compute_mvmZ(SGVector<float64_t> x)
+{
+	//time complexity O(m*n)
+	Map<MatrixXd> eigen_Rvdd(m_Rvdd.matrix, m_Rvdd.num_rows, m_Rvdd.num_cols);
+	Map<VectorXd> eigen_t(m_t.vector, m_t.vlen);
+	Map<VectorXd> eigen_x(x.vector, x.vlen);
+
+	SGVector<float64_t> res(x.vlen);
+	Map<VectorXd> eigen_res(res.vector, res.vlen);
+
+	//Zx = t.*x - RVdd'*(RVdd*x);
+	eigen_res=eigen_x.cwiseProduct(eigen_t)-eigen_Rvdd.transpose()*(eigen_Rvdd*eigen_x);
+	return res;
+}
+
+SGVector<float64_t> CSingleFITCLaplacianInferenceMethod::compute_mvmK(SGVector<float64_t> al)
+{
+	//time complexity O(m*n)
+	Map<MatrixXd> eigen_V(m_V.matrix, m_V.num_rows, m_V.num_cols);
+	Map<VectorXd> eigen_dg(m_dg.vector, m_dg.vlen);
+	Map<VectorXd> eigen_al(al.vector, al.vlen);
+
+	SGVector<float64_t> res(al.vlen);
+	Map<VectorXd> eigen_res(res.vector, res.vlen);
+
+	//Kal = V'*(V*al) + d0.*al;
+	eigen_res= eigen_V.transpose()*(eigen_V*eigen_al)+eigen_dg.cwiseProduct(eigen_al); 
+	return res;
+}
+
+CSingleFITCLaplacianInferenceMethod* CSingleFITCLaplacianInferenceMethod::obtain_from_generic(
+		CInferenceMethod* inference)
+{
+	REQUIRE(inference!=NULL, "Inference should be not NULL");
+
+	if (inference->get_inference_type()!=INF_FITC_LAPLACIAN_SINGLE)
+		SG_SERROR("Provided inference is not of type CSingleFITCLaplacianInferenceMethod!\n")
+
+	SG_REF(inference);
+	return (CSingleFITCLaplacianInferenceMethod*)inference;
+}
+
+SGMatrix<float64_t> CSingleFITCLaplacianInferenceMethod::get_chol_inv(SGMatrix<float64_t> mtx)
+{
+	//time complexity O(m^3), where mtx is a m-by-m matrix
+	REQUIRE(mtx.num_rows==mtx.num_cols, "Matrix must be square\n");
+
+	Map<MatrixXd> eigen_mtx(mtx.matrix, mtx.num_rows, mtx.num_cols);
+	LLT<MatrixXd> chol(eigen_mtx.colwise().reverse().rowwise().reverse().matrix());
+	//tmp=chol(rot180(A))'
+	MatrixXd tmp=chol.matrixL();
+	SGMatrix<float64_t> res(mtx.num_rows, mtx.num_cols);
+	Map<MatrixXd> eigen_res(res.matrix, res.num_rows, res.num_cols);
+	//chol_inv = @(A) rot180(chol(rot180(A))')\eye(nu);                 % chol(inv(A))
+	eigen_res=tmp.colwise().reverse().rowwise().reverse().matrix().triangularView<Upper>(
+		).solve(MatrixXd::Identity(mtx.num_rows, mtx.num_cols));
+	return res;
+}
+
+float64_t CSingleFITCLaplacianInferenceMethod::get_negative_log_marginal_likelihood()
+{
+	if (parameter_hash_changed())
+		update();
+
+	if (m_Wneg)
+	{
+		SG_WARNING("nlZ cannot be computed since W is too negative");
```

Comment:
too negative? Maybe some details how to avoid? Keep in mind users dont know what W is